[JENKINS-33296] - plugin dependency issues during install

kzantow opened the pull request:
This addresses a few issues with dependent plugins:
- previously "implied" plugin dependencies are not available without explicit dependency references
- maven-plugin was being explicitly excluded as a dependency, causing many failures
- mailer had explicit dependencies in the main codebase for functionality, including saving the 'email' address field for Jenkins user accounts
- dependencies were included in client-side search text in the installer

This addresses:
https://issues.jenkins-ci.org/browse/JENKINS-33296
https://issues.jenkins-ci.org/browse/JENKINS-33308

@jenkinsci/code-reviewers

kzantow committed:
JENKINS-33296 - plugin dependency issues during install

daniel-beck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -379,19 +379,31 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
+    public boolean isMailerPluginPresent() {
+        try {
+            // mail support has moved to a separate plugin
+            return null != Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
```
Wouldn't it be easier to query the plugin manager for `mailer`? But probably safer this way…

slide replied:
I wish I had never moved this UserProperty out of core...

jglick replied:
:ant: `loadClass` cannot ever return null, so you need not and should not inspect the return value.

```java
try {
    Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
    return true;
} catch (ClassNotFoundException e) {
    return false;
}
```

jglick replied:
> I wish I had never moved this UserProperty out of core...

Creating files like `Mailer$UserProperty.java` is technically legal, but pretty messy! Anyway might cause linkage errors now (mismatch in outer class code attribute; cannot remember if this applies only to inner classes or also to nested classes).

kzantow replied:
@jglick right, most of the changes here were just from refactoring

daniel-beck commented:
For reference, while this fixes Maven Plugin related issues, it does not take care of [JENKINS-33308](https://issues.jenkins-ci.org/browse/JENKINS-33308) type dependency problems.

kzantow commented:
@daniel-beck right, it's only a part of the issue

kzantow committed:
JENKINS-33308 - automatically install previously bundled dependencies
for plugins that depend on old Jenkins versions

kzantow commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1140,7 +1140,9 @@ public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
-                    // TODO: Someone that understands what the following logic is about, please add a comment.
+                    // There could be cases like:
+                    // 'plugin.ambiguous.updatesite' where both
+                    // 'plugin' @ 'ambigiuous.updatesite' and 'plugin.ambiguous' @ 'updatesite' resolve to valid plugins
```
Unrelated, but no need for a TODO here

kwhetstone replied:
Hey, adding doc is always welcome.

kzantow commented:
@reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

jglick commented on war/src/main/js/api/plugins.js:
```diff
@@ -19,7 +19,7 @@ exports.recommendedPlugins = [
     "ldap",
     "mailer",
-    // "matrix-auth",
+    "matrix-auth",
```
Effectively reverting 7f87486c73b1c63cc595c0fc380d42ada45f6116 I suppose.

kzantow replied:
Yes, since the issues are 'fixed'

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -259,6 +259,19 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
+    /**
+     * Returns all the bundled plugin dependencies for a particular Jenkins version
+     */
+    public static List<PluginWrapper.Dependency> getPreviouslyBundledDependencies(String jenkinsVersion) {
```
:ant: the naming is misleading. Whether these were _bundled_ in 1.x or not is irrelevant—only whether they were _detached_ from core. (There are some bundled plugins which were never detached, and we could have unbundled detached plugins though as a matter of policy I think we never did.)

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -259,6 +259,19 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
+        List<PluginWrapper.Dependency> out = new ArrayList<>();
+        for (DetachedPlugin detached : DETACHED_LIST) {
+            if (jenkinsVersion == null || jenkinsVersion.equals("null") || new VersionNumber(jenkinsVersion).compareTo(detached.splitWhen) <= 0) {
+                out.add(new PluginWrapper.Dependency(detached.shortName + ':' + detached.requireVersion));
```
:bug: reuse `DetachedPlugin.fix`, not only to avoid duplication but so you can honor `BREAK_CYCLES`.

jglick replied:
(probably after refactoring that method to accept `String yourName, String jenkinsVersion` rather than `Attributes atts`)

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1150,9 +1152,23 @@ public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
             }
+            
+            // JENKINS-33308 - automatically install implied/previously bundled dependencies for older plugins that may need them
+            for(PluginWrapper.Dependency previouslyBundledDependency : ClassicPluginStrategy.getPreviouslyBundledDependencies(p.requiredCore)) {
```
This smells like the wrong place to be doing this. The detached plugin mapping should be logically applied to incoming metadata, whether from UC JSON or from a plugin manifest. You seem to be adding in these dependencies at the last minute, whereas they should been considered dependencies from the start.

I suspect that the difference is tangible: if you go into the setup wizard and check the box next to some ancient plugin, the box for the detached dependency like `matrix-project` should get checked then too (which of course _also_ means that we will wind up installing it, but without any special handling _here_).

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -634,10 +634,8 @@ public Plugin(String sourceId, JSONObject o) {
             this.categories = o.has("labels") ? (String[])o.getJSONArray("labels").toArray(new String[0]) : null;
             for(Object jo : o.getJSONArray("dependencies")) {
                 JSONObject depObj = (JSONObject) jo;
-                // Make sure there's a name attribute, that that name isn't maven-plugin - we ignore that one -
```
WTF

jglick commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -379,19 +379,31 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
-            throw e;
-        } catch (Exception e) {
-            LOGGER.log(Level.WARNING, "Failed to set the e-mail address",e);
+        if(isMailerPluginPresent()) {
```
Simpler to just catch, and ignore, `ClassNotFoundException` right here.

jglick commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -379,19 +379,31 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
         return user;
     }
+    
+    public boolean isMailerPluginPresent() {
```
:bug: should not be an API:

```java
@Restricted(NoExternalUse.class) // just here for Jelly
```

jglick commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -599,7 +599,13 @@ var createPluginSetupWizard = function(appendTarget) {
+					if(localMatches.length > 0) {
+						matches = matches.concat(localMatches);
+					}
+				});
```
Is this related?

kzantow replied:
Related to issues with dependencies, insomuch as they were erroneously being included in search results in the installer

jglick commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -358,7 +358,7 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
             si.fullname = si.username;
 
-        if(si.email==null || !si.email.contains("@"))
+        if(isMailerPluginPresent() && (si.email==null || !si.email.contains("@")))
```
This is really crying out for an extension point—a way for `UserPropertyDescriptor` to indicate that it wishes to be configured during account creation, for security realms supporting such a concept.

kzantow replied:
Agree, but given the timing and existing code, I think handling of this is probably sufficient: this at least hides the fields that would not save properly, which would be very confusing and buggy to the user

jglick replied:
Yes I think this is good enough for now, just mentioning it for the record.

daniel-beck replied:
> This is really crying out for an extension point

Agreed, would be great, but out of scope for this PR and 2.0 I think. 2.1 maybe…

kzantow committed:
Address @jglick's comments: refactor ClassicPluginStrategy to use
BREAK_CYCLE, use the same getImpliedDependency method for multiple code
paths, @Restricted isMailerPresent

kzantow committed:
Implement @jglick's idea: move implicit dependency modifications to
UpdateSite, when reading plugin metadata

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -612,7 +623,7 @@ public Api getApi() {
 
         /**
-         * Dependencies of this plugin.
+         * Dependencies of this plugin, a name -&gt; version mapping.
```
BTW `→` requires no escape. :-)

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -260,6 +260,40 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
     }
 
+    private static void fix(Attributes atts, List<PluginWrapper.Dependency> optionalDependencies) {
+        // don't fix the dependency for yourself, or else we'll have a cycle
```
This comment might make more sense inside `getImpliedDependencies` where the `pluginName` is used.

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -260,6 +260,40 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
+    
+    /**
+     * Returns all the plugin dependencies that are implicit based on a particular Jenkins version
+     */
```
`@since TODO`

jglick replied:
(assuming you expect this to be public)

jglick replied:
(BTW for `plugin-compat-tester` we would actually need the list to be present in a static resource. CC @kwhetstone.)

kzantow replied:
Is it fine to leave it literally "@since TODO" should I guess at the upcoming version? Sorry, just don't know what to do here, specifically.

jglick replied:
As a rule in PRs we just leave it as `TODO`, with the expectation that the person merging the PR should replace that with the actual version then in the POM. Of course in this case you expect the version to be 2.0 (or some uninteresting alpha variant thereof).

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -467,7 +468,17 @@ public boolean isLegacyDefault() {
+                if(!implicitDeps.isEmpty()) {
+                    for(PluginWrapper.Dependency dep : implicitDeps) {
+                        if(!p.dependencies.containsKey(dep.shortName)) {
+                            p.dependencies.put(dep.shortName, dep.version);
```
FWIW in `createPluginWrapper` these are added as optional dependencies—I am not really sure why, and that is perhaps part of the problem in JENKINS-33308.

kzantow replied:
Yes, they're added as optional dependencies, but this is too late in the process, I believe this just makes the classes available at runtime?

jglick commented:
:bee: but please check behavior w.r.t. disabling implied dependencies which I suspect is still broken. (For example with the Flexible Publish → Matrix Project dep.)

kzantow commented:
I did test the update center attempting to remove an implied dependency, it treats them as normal, required dependencies and doesn't let you. @jglick

kzantow committed:
Refactoring lack-of-moving-comments-too fail

jglick commented:
> disabling implied dependencies which I suspect is still broken

Yes…so you have fixed the *UI* portion of the code to consider these, and did so as mandatory dependencies. I suspect that the mistake in `PluginManager` (i.e., the non-UI part) treating them as optional will not become visible until we have something like #2001. You would expect for example that

    touch $JENKINS_HOME/work/plugins/matrix-project.jpi.disabled

and restarting would cause Flexible Publish to be disabled, too. Even if there were such a defense in core—today there is not—the inexplicable use of `optionalDependencies` for detached dependencies means that it would still not work for these cases.

Thus not a regression in your PR, but rather an apparent bug in code your PR refactored.

daniel-beck commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -260,6 +260,41 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
                 createClassLoader(paths, dependencyLoader, atts), disableFile, dependencies, optionalDependencies);
     }
 
+    private static void fix(Attributes atts, List<PluginWrapper.Dependency> optionalDependencies) {
```
Moving code around makes diffs difficult to read.

daniel-beck commented:
From reading the code, this looks like :+1: to me. Now on to the testing…

kwhetstone commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -379,19 +378,30 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
+                user.addProperty((UserProperty)c.newInstance(si.email));
+            } catch (ReflectiveOperationException e) {
+                throw new RuntimeException(e);
+            }
```
Maybe log a message here with the attempted action.

kzantow replied:
It's not attempting anything, it doesn't proceed into this block unless isMailerPluginPresent() is true, in which case the class check is already done and instead of silently failing as it was doing, it will throw an exception.

kwhetstone commented:
:bee:

daniel-beck commented:
Seems pretty solid in testing. It's annoying as hell, but that's how implied dependencies work…
